src/ComplexType.php

```diff
@@ -101,93 +120,228 @@ protected function generateClass()
-                        . '  }' . PHP_EOL;
-                } else {
-                    $setterCode = '  $this->' . $name . ' = $' . $name . '->format(\DateTime::ATOM);' . PHP_EOL;
-                }
-            } else {
-                $setterCode = '  $this->' . $name . ' = $' . $name . ';' . PHP_EOL;
-            }
-            $setterCode .= '  return $this;' . PHP_EOL;
-            $setter = new PhpFunction(
-                'public',
-                'set' . ucfirst($name),
-                $this->buildParametersString(
-                    array($name => $typeHint),
-                    true,
-                    // If the type of a member is nullable we should allow passing null to the setter. If the type
-                    // of the member is a class and not a primitive this is only possible if setter parameter has
-                    // a default null value. We can detect whether the type is a class by checking the type hint.
-                    $member->getNullable() && !empty($typeHint)
-                ),
-                $setterCode,
-                $setterComment
+            $this->addProperty($name, $type);
+            $this->addGetter($name, $type);
+            $this->addSetter($name, $type, $typeHint, $nullable);
+        }
+
+        $constructor->setBody($constructorSource);
+    }
+
+    /**
+     * Add parameter to method
+     *
+     * @param MethodGenerator $method
+     * @param $name
+     * @param $type
+     * @param $typeHint
+     * @param $defaultToNull
+     */
+    private static function addMethodParameter(MethodGenerator $method, $name, $type, $typeHint, $defaultToNull)
+    {
+        $parameter = new ParameterGenerator();
+        $parameter->setName($name);
+
+        $method->getDocBlock()->setTag(new ParamTag($name, $type));
+
+        if (!empty($typeHint)) {
+            $parameter->setType($typeHint);
+        }
+        if ($defaultToNull) {
+            $parameter->setDefaultValue(
+                new ValueGenerator(null, ValueGenerator::TYPE_NULL)
             );
-            $accessors[] = $setter;
         }
+        $method->setParameter($parameter);
+    }
+
+    /**
+     * Add inherited parameter to constructor
+     *
+     * @param MethodGenerator $constructor
+     * @param $name
+     * @param $type
+     * @param $typeHint
+     * @param $constructorParamsDefaultToNull
+     * @return ParameterGenerator
+     */
+    private static function addParentConstructorParameter(
+        MethodGenerator $constructor,
+        $name, $type, $typeHint, $constructorParamsDefaultToNull
+    )
+    {
+        self::addMethodParameter($constructor, $name, $type, $typeHint, $constructorParamsDefaultToNull);
+
+        $parentConstructorParameter = new ParameterGenerator();
+        $parentConstructorParameter->setName($name);
+        if (!empty($typeHint)) {
+            $parentConstructorParameter->setType($typeHint);
+        }
```

Comment:
`addMethodParameter()` already builds a `ParameterGenerator`. Could we return this instead of building a new one?

They seem very similar except this one does not have a default value set.